test/tokens/TokenManager.js

```diff
@@ -0,0 +1,133 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/* global describe it */
+'use strict';
+
+const assert = require('assert');
+
+const TokenManager = require('../../lib/tokens/TokenManager');
+
+class TestTokenManager1 extends TokenManager {
+  constructor() {
+    super();
+    this._getTokenCallCount = 0;
+  }
+
+  _getToken(callback) {
+    this._getTokenCallCount += 1;
+    setTimeout(() => {
+      callback(null, { headers: { 'set-cookie': [ 'Max-Age=1' ] } });
+    }, 100);
+  }
+
+  // mock successful token renewal
+  get getTokenCallCount() {
+    return this._getTokenCallCount;
+  }
+}
+
+class TestTokenManager2 extends TokenManager {
+  constructor() {
+    super();
+    this._getTokenCallCount = 0;
+  }
+
+  // mock failed token renewal
+  _getToken(callback) {
+    this._getTokenCallCount += 1;
+    setTimeout(() => {
+      callback(new Error('err'), {ok: false});
+    }, 100);
+  }
+
+  get getTokenCallCount() {
+    return this._getTokenCallCount;
+  }
+}
+
+describe('Token Manger', (done) => {
+  it('renews the token successfully', (done) => {
+    let t = new TestTokenManager1();
+    t.renew().then(() => {
+      assert.equal(t.getTokenCallCount, 1);
+      done();
+    }).catch(done);
+    assert.ok(t.getRenewingToken);
+  });
+
+  it('handles a token renewal failure', (done) => {
+    let t = new TestTokenManager2();
+    t.renew().then(() => {
+      assert.fail('Unexpected success.');
+    }).catch((error) => {
+      assert.equal(t.getTokenCallCount, 1);
+      assert.equal(error.message, 'err');
+      assert.equal(error.response.ok, false);
+      done();
+    });
+    assert.ok(t.getRenewingToken);
+  });
+
+  it('correctly auto renews token', (done) => {
+    let t = new TestTokenManager1();
+    t.startAutoRenew();
+    setTimeout(() => {
+      // one renew every 0.5 seconds
+      assert.equal(t.getTokenCallCount, 4);
+      done();
+    }, 2000);
+  });
+
+  it('only makes one renewal request', (done) => {
+    let t = new TestTokenManager1();
+    let renewalCount = 0;
+    let lim = 10000;
+
+    for (let i = 1; i < lim + 1; i++) {
+      if (i === lim) {
+        t.renew().then(() => {
+          renewalCount += 1;
+          assert.equal(renewalCount, lim);
+          assert.equal(t.getTokenCallCount, 1);
+          done();
+        }).catch(done);
+      } else {
+        t.renew().then(() => {
+          renewalCount += 1;
+        }).catch(done);
+      }
+    }
+    assert.ok(t.getRenewingToken);
+  });
+
+  it('makes another renewal only after setting force renew', (done) => {
+    let t = new TestTokenManager1();
+    // renew 1 - make request
+    t.renew().then(() => {
+      assert.equal(t.getTokenCallCount, 1);
+      // renew 2 - return last good response
+      t.renew().then(() => {
+        assert.equal(t.getTokenCallCount, 1);
```

Comment:
Should we also assert a mocked `_lastRenewResponse`?

Reply:
See above comment. I'll remove the `_lastRenewResponse` thing.